### PR TITLE
Feature make oauth `audience` configurable

### PR DESCRIPTION
**Pull request by himadripal**
[Make the OAuth2 request audience
](https://github.com/apache/iceberg-python/issues/479)

### CODE REVIEW

**Comment by himadripal** [comment]
Although the issue discusses about making `audience` configurable but there are few other optional parameter exists in the specification (I have added support for `resource` as an additional example) which are not configurable right now. This is an attempt to make all optional OAuth params passing configurable. Have introduced a map instead of single param.  Idea is if you have those optional parameters configured in properties, we build map out of it, picking only optional params from the properties and pass it along. 
`scope` is also an optional parameter, I think it can also be moved to using this map.
 @nastra @rdblue @syun64 Please take a look and provide feedback.

**Comment by himadripal** [comment]
Can I get someone's attention for a review on this one please?

**Comment by flyrain** [comment]
Thanks @himadripal for working on it. Let me take a look.

**Review by flyrain — COMMENTED** [review]
Thanks @himadripal working on it. LGTM overall. Left some comments.

File: core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java [hunk 1]
```diff
@@ -178,15 +178,18 @@ public void initialize(String name, Map<String, String> unresolved) {
     ConfigResponse config;
     OAuthTokenResponse authResponse;
     String credential = props.get(OAuth2Properties.CREDENTIAL);
+    // CONSIDER : putting scope in optional param map - reduce wiring on scope
```
[the remark this answers is not in the document]
Comment (himadripal) [reply core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:181]: >Ideally, we should have a Java class synced with the [OAuthClientCredentialsRequest](https://github.com/apache/iceberg/blob/bcbcbb263ea7e13ab22d0feb918e207c7e42dbbd/open-api/rest-catalog-open-api.yaml#L2919) in Rest spec

A java class for all optional params is good suggestion, that way, we can implement individual parameter level default and validation if we need any in future.

> However, that's fairly a big change. We will potentially need to deprecate some methods. Another option is to put every optional parameter(including scope, token type) in a map at this PR, then refactor it later.

Yes, this will need changes in public API and deprecate some methods. I tried putting `scope` in the map and some of the tests were failing, I'll take a look again. How about we put all the optional param mentioned [here](https://datatracker.ietf.org/doc/html/rfc8693#name-token-exchange-request-and-) in the map and leave the scope handling for later ( I'll create an issue for consolidating this).

**Review by flyrain — APPROVED** [review]
+1. Thanks @himadripal for working on it. As a followup, would you mind create issues?
1. Adding audience, resource in the rest spec.
2. Enriching Rest catalog configure doc with more options, in spark-configuration.md, we usually use the spark as the reference implementation, so it's OK to only change the spark side.

File: core/src/test/java/org/apache/iceberg/rest/TestRESTCatalog.java [hunk 2]
```diff
@@ -1444,7 +1484,7 @@ public void testCatalogTokenRefresh(String oauth2ServerUri) {
                       eq(HTTPMethod.POST),
                       eq(oauth2ServerUri),
                       any(),
-                      Mockito.argThat(firstRefreshRequest::equals),
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/rest/TestRESTCatalog.java:1447]: unrelated change

File: core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java [hunk 3]
```diff
@@ -471,7 +515,13 @@ static void setTokenRefreshNumRetries(int retries) {
...
     public static AuthSession empty() {
       return new AuthSession(
-          ImmutableMap.of(), null, null, null, OAuth2Properties.CATALOG_SCOPE, null);
+          ImmutableMap.of(),
+          null,
+          null,
+          null,
+          OAuth2Properties.CATALOG_SCOPE,
+          null,
+          ImmutableMap.of());
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java:524]: it has become quite painful to add new functionality/parameters to `AuthSession`, since it requires adding a single param at a bunch of places. The most recent change was done by #8976.

A while ago I experimented with an [`AuthConfig`](https://github.com/apache/iceberg/commit/3bed59907f4230d10a97742c58ab4b85ce7b1ebf#diff-d783470ff004c360395125d467f1aadd776c8c9e925ba52639a405a03a72b870R647), which would carry all parameters required for an `AuthSession`, so that it would be easier to add new configuration parameters to it. So I wonder if we should refactor `AuthSession` and its configuration so that it's less painful to add new stuff to it?

/cc @danielcweeks @amogh-jahagirdar @rdblue

Comment (himadripal) [reply core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java:524]: +1 to the painful to add parameters. With this PR, I have tried to add all optional oauth parameters (including scope) in a map but having a class like `[AuthConfig](https://github.com/apache/iceberg/commit/3bed59907f4230d10a97742c58ab4b85ce7b1ebf#diff-d783470ff004c360395125d467f1aadd776c8c9e925ba52639a405a03a72b870R647),` which holds all params with validation logic and passing it around would be best.

**Comment by nastra** [comment]
overall I think adding these optional OAuth2 params make sense to me. I'll also link to the [OAuth2 spec](https://datatracker.ietf.org/doc/html/rfc8693#name-token-exchange-request-and-) for other reviewers to better understand what those optional parameters do.
I think we should decide how we want to deal with `AuthSession` configuration paramters as it causes lots of changes to add a single parameter to `AuthSession`.

File: core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java [hunk 4]
```diff
@@ -257,8 +292,9 @@ private static Map<String, String> tokenExchangeRequest(
       formData.put(ACTOR_TOKEN, actorToken);
       formData.put(ACTOR_TOKEN_TYPE, actorTokenType);
     }
+    formData.putAll(optionalParams);
 
-    return formData.build();
+    return formData.buildKeepingLast();
```
Comment (himadripal) [inline core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java:297]: please note, changed it to `buildKeepingLast()` , I noticed everywhere else `build` is used which throws exception if it encounters duplicate keys.

File: core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java (hunk 3 above)
Comment (flyrain) [reply core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java:529]: Can we generate the request class(e.g., [OAuthClientCredentialsRequest](https://github.com/apache/iceberg/blob/bcbcbb263ea7e13ab22d0feb918e207c7e42dbbd/open-api/rest-catalog-open-api.yaml#L2919) ) from the spec so that we can pass a request object with all required and optional params? By doing that, we will always make sure the impl. synced with spec.

**Comment by flyrain** [comment]
+1 Thanks @himadripal for the PR. Thanks @nastra for the review.
